br/pkg/lightning/backend/local/duplicate.go

```diff
@@ -573,6 +590,122 @@ func (m *DupeDetector) RecordIndexConflictError(ctx context.Context, stream DupK
 	return nil
 }
 
+// RetrieveKeyAndValueFromErrFoundDuplicateKeys retrieves the key and value
+// from ErrFoundDuplicateKeys error.
+func RetrieveKeyAndValueFromErrFoundDuplicateKeys(err error) ([]byte, []byte, error) {
+	if !common.ErrFoundDuplicateKeys.Equal(err) {
+		return nil, nil, err
+	}
+	tErr, ok := errors.Cause(err).(*terror.Error)
+	if !ok {
+		return nil, nil, err
+	}
+	if len(tErr.Args()) != 2 {
+		return nil, nil, err
+	}
+	key, keyIsByte := tErr.Args()[0].([]byte)
+	value, valIsByte := tErr.Args()[1].([]byte)
+	if !keyIsByte || !valIsByte {
+		return nil, nil, err
+	}
+	return key, value, nil
+}
+
+// NewErrFoundConflictRecords generate an error ErrFoundDataConflictRecords / ErrFoundIndexConflictRecords
+// according to key and value.
+func NewErrFoundConflictRecords(key []byte, value []byte, tbl table.Table, idxInfo *model.IndexInfo) error {
+	sessionOpts := encode.SessionOptions{
+		SQLMode: mysql.ModeStrictAllTables,
+	}
+
+	decoder, err := kv.NewTableKVDecoder(tbl, tbl.Meta().Name.L, &sessionOpts, log.L())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if tablecodec.IsRecordKey(key) {
+		// for data KV
+		handle, err := tablecodec.DecodeRowKey(key)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		rowData := decoder.DecodeRawRowDataAsStr(handle, value)
+
+		return errors.Trace(common.ErrFoundDataConflictRecords.FastGenByArgs(tbl.Meta().Name, handle.String(), rowData))
+	}
+
+	// for index KV
+	return NewErrFoundIndexConflictRecords(key, value, tbl, idxInfo)
+}
+
+// NewErrFoundIndexConflictRecords generate an error ErrFoundIndexConflictRecords
+// according to key and value.
+func NewErrFoundIndexConflictRecords(key []byte, value []byte, tbl table.Table, idxInfo *model.IndexInfo) error {
+	sessionOpts := encode.SessionOptions{
+		SQLMode: mysql.ModeStrictAllTables,
+	}
+
+	decoder, err := kv.NewTableKVDecoder(tbl, tbl.Meta().Name.L, &sessionOpts, log.L())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	_, idxID, _, err := tablecodec.DecodeIndexKey(key)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if idxInfo == nil {
```

Comment:
can we restrict that idxInfo must not be nil? I see a possible nil path from ConvertToErrFoundConflictRecords -> NewErrFoundConflictRecords -> NewErrFoundIndexConflictRecords. We can let NewErrFoundConflictRecords rather than NewErrFoundIndexConflictRecords find the idxInfo